gprofiler/proc_events.py

```diff
@@ -0,0 +1,179 @@
+import os
+import selectors
+import socket
+import struct
+import threading
+
+
+def _raise_if_not_running(func):
+    def wrapper(self, *args, **kwargs):
+        if not self.is_alive():
+            raise RuntimeError("Process Events Listener wasn't started")
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class _ProcEventsListener(threading.Thread):
+    """Thead listening to process events.
+
+    Notice that there cannot be more than a single instance of this class per-process because it opens a
+    process-connector socket that can (probably) not be opened twice in the same process.
+    """
+
+    # linux/netlink.h:
+    _NETLINK_CONNECTOR = 11
+
+    # linux/netlink.h:
+    _NLMSG_DONE = 0x3  # End of a dump
+
+    # struct nlmsghdr {
+    #         __u32           nlmsg_len;      /* Length of message including header */
+    #         __u16           nlmsg_type;     /* Message content */
+    #         __u16           nlmsg_flags;    /* Additional flags */
+    #         __u32           nlmsg_seq;      /* Sequence number */
+    #         __u32           nlmsg_pid;      /* Sending process port ID */
+    # };
+    _nlmsghdr = struct.Struct("=I2H2I")
+
+    # linux/connector.h:
+    _CN_IDX_PROC = 0x1
+    _CN_VAL_PROC = 0x1
+
+    # struct cn_msg {
+    #         struct cb_id id;
+    #
+    #         __u32 seq;
+    #         __u32 ack;
+    #
+    #         __u16 len;              /* Length of the following data */
+    #         __u16 flags;
+    #         __u8 data[0];
+    # };
+    _cn_msg = struct.Struct("=4I2H")
+
+    # linux/cn_proc.h:
+    # struct proc_event {
+    #         enum what {
+    #                 ...
+    #         } what;
+    #         __u32 cpu;
+    #         __u64 __attribute__((aligned(8))) timestamp_ns;
+    #                 /* Number of nano seconds since system boot */
+    #         union { /* must be last field of proc_event struct */
+    #                 ...
+    #         } event_data;
+    # };
+    _base_proc_event = struct.Struct("=2IQ")
+
+    # From enum what
+    _PROC_EVENT_EXIT = 0x80000000
+
+    # From enum proc_cn_mcast_op
+    _PROC_CN_MCAST_LISTEN = 1
+
+    # struct exit_proc_event {
+    #         __kernel_pid_t process_pid;
+    #         __kernel_pid_t process_tgid;
+    #         __u32 exit_code, exit_signal;
+    # } exit;
+    _exit_proc_event = struct.Struct("=4I")
+
+    def __init__(self):
+        self._socket = socket.socket(socket.AF_NETLINK, socket.SOCK_DGRAM, self._NETLINK_CONNECTOR)
+        self._exit_callbacks = []
+        self._should_stop = False
+
+        self._selector = selectors.DefaultSelector()
+        # Create a pip so we can make select() return
+        r, w = os.pipe()
+        self._select_breaker = os.fdopen(w, "w")
```

Comment:
Just making sure - Python's `fdopen` does not `dup` or anything, right? So `r, w` are not leaked

Reply:
If it does it would've been a bug.
On the other hand, I never close the pipe files, which I should.